python/repositories.bzl

```diff
@@ -1,8 +1,252 @@
+            )
+            zstd = "{working_directory}/zstd".format(working_directory = working_directory)
+            unzstd = "./unzstd"
+            rctx.symlink(zstd, unzstd)
+
+        exec_result = rctx.execute([
+            "tar",
+            "--extract",
+            "--strip-components=2",
+            "--use-compress-program={unzstd}".format(unzstd = unzstd),
+            "--file={}".format(release_filename),
+        ])
+        if exec_result.return_code:
+            fail(exec_result.stderr)
+    else:
+        rctx.download_and_extract(
+            url = url,
+            sha256 = rctx.attr.sha256,
+            stripPrefix = "python",
+        )
+
+    # Make the Python installation read-only.
+    if "windows" not in rctx.os.name:
+        exec_result = rctx.execute(["chmod", "-R", "ugo-w", "lib"])
+        if exec_result.return_code:
+            fail(exec_result.stderr)
+
+    python_bin = "python.exe" if ("windows" in rctx.attr.platform) else "bin/python3"
+
+    build_content = """\
+# Generated by python/repositories.bzl
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "files",
+    srcs = glob(
+        include = [
+            "bin/**",
+            "DLLs/**",
+            "extensions/**",
+            "include/**",
+            "lib/**",
+            "libs/**",
+            "Scripts/**",
+            "share/**",
+        ],
+        exclude = [
+            "**/* *", # Bazel does not support spaces in file names.
+        ],
+    ),
+)
+
+exports_files(["{python_path}"])
+
+py_runtime(
+    name = "py2_runtime",
+    interpreter_path = "/bin/false",
+    python_version = "PY2",
+)
+
+py_runtime(
+    name = "py3_runtime",
+    files = [":files"],
+    interpreter = "{python_path}",
+    python_version = "PY3",
+)
+
+py_runtime_pair(
+    name = "python_runtimes",
+    py2_runtime = ":py2_runtime",
+    py3_runtime = ":py3_runtime",
+)
+""".format(
+        python_path = python_bin,
+    )
+    rctx.file("BUILD.bazel", build_content)
+
+    return {
+        "name": rctx.attr.name,
+        "platform": platform,
+        "python_version": python_version,
+        "sha256": rctx.attr.sha256,
+    }
+
+python_repository = repository_rule(
+    _python_repository_impl,
+    doc = "Fetches the external tools needed for the Python toolchain.",
+    attrs = {
+        "platform": attr.string(
+            doc = "The platform name for the Python interpreter tarball.",
+            mandatory = True,
+            values = PLATFORMS.keys(),
+        ),
+        "python_version": attr.string(
+            doc = "The Python version.",
+            mandatory = True,
+            values = TOOL_VERSIONS.keys() + MINOR_MAPPING.keys(),
+        ),
+        "sha256": attr.string(
+            doc = "The SHA256 integrity hash for the Python interpreter tarball.",
+            mandatory = True,
```

Comment:
I don't think this should be mandatory. Allowing users to not set it will cause bazel to post the usual logging informing users they can improve their experience by filling this in. My general workflow when adding new dependencies is to add the url, do a sync, then fill in the blanks based on the warnings.

Reply:
I created a tool to get the hashes here: https://github.com/bazelbuild/rules_python/pull/618/commits/d7ac042f2ad92670f60e25d4f97fcdef567c7859.